File: BFF_Spletni/routes/movies.js
const Router = require("koa-router");
const client = require("../client");
const jwtMiddleware = require("../authMiddleware");
const roleMiddleware = require("../roleMiddleware");


const router = new Router({
  prefix: "/movies",
});

// Create movie
router.post("/", jwtMiddleware, async (ctx) => {
  const data = ctx.request.body;
  await new Promise((resolve, reject) => {
    client.CreateMovie(data, (err, res) => {
      if (err) {
        ctx.status = 500;
        ctx.body = { error: err.message };
        return reject(err);
      }
      ctx.body = res;
      resolve();
    });
  });
});

// Get movie by ID
router.get("/:id", async (ctx) => {
  const id = parseInt(ctx.params.id, 10);
  await new Promise((resolve, reject) => {
    client.GetMovieById({ id }, (err, res) => {
      if (err) {
        ctx.status = 404;
        ctx.body = { error: err.message };
        return reject(err);
      }
      ctx.body = res;
      resolve();
    });
  });
});

// Search movies
router.get("/", async (ctx) => {
  const query = ctx.query.q;
  if (!query) {
    ctx.status = 400;
    ctx.body = { error: "Missing search query (?q=)" };
    return;
  }

  await new Promise((resolve, reject) => {
    client.SearchMovies({ query }, (err, res) => {
      if (err) {
        ctx.status = 500;
        ctx.body = { error: err.message };
        return reject(err);
      }
      ctx.body = res.movies;
      resolve();
    });
  });
});

// Update movie
router.put("/:id", jwtMiddleware, roleMiddleware, async (ctx) => {
  const id = parseInt(ctx.params.id, 10);
  const data = { ...ctx.request.body, id };

  await new Promise((resolve, reject) => {
    client.UpdateMovie(data, (err, res) => {
      if (err) {
        ctx.status = 500;
        ctx.body = { error: err.message };
        return reject(err);
      }
      ctx.body = res;
      resolve();
    });
  });
});

// Delete movie
router.delete("/:id", jwtMiddleware, roleMiddleware, async (ctx) => {
  const id = parseInt(ctx.params.id, 10);
  await new Promise((resolve, reject) => {
    client.DeleteMovie({ id }, (err, res) => {
      if (err) {
        ctx.status = 500;
        ctx.body = { error: err.message };
        return reject(err);
      }
      ctx.body = res;
      resolve();
    });
  });
});

module.exports = router;

File: BFF_Mobilni/routes/movies.py
import grpc
import movies_pb2
import movies_pb2_grpc
from flask import Blueprint, jsonify, request

# Set up the gRPC channel and client
channel = grpc.insecure_channel('localhost:50051')
movie_stub = movies_pb2_grpc.MovieServiceStub(channel)

# Create a Blueprint for movie-related routes
movie_routes = Blueprint('movie_routes', __name__)

# Route to create a movie via gRPC
@movie_routes.route('/', methods=['POST'])
def create_movie():
    data = request.get_json()
    movie_request = movies_pb2.MovieCreateRequest(
        title=data.get('title'),
        year=data.get('year'),
        genre=data.get('genre'),
        description=data.get('description'),
        actors=data.get('actors', [])
    )
    response = movie_stub.CreateMovie(movie_request)
    return jsonify({
        "id": response.id,
        "title": response.title,
        "year": response.year,
        "genre": response.genre,
        "description": response.description,
        "actors": list(response.actors)  # Convert to Python list
    }), 201

# Route to get movie by ID via gRPC
@movie_routes.route('/<int:id>', methods=['GET'])
def get_movie_by_id(id):
    movie_request = movies_pb2.MovieRequest(id=id)
    response = movie_stub.GetMovieById(movie_request)
    return jsonify({
        "id": response.id,
        "title": response.title,
        "year": response.year,
        "genre": response.genre,
        "description": response.description,
        "actors": list(response.actors)  # Convert to Python list
    }), 200

# Route to update a movie via gRPC
@movie_routes.route('/<int:id>', methods=['PUT'])
def update_movie(id):
    data = request.get_json()
    movie_update_request = movies_pb2.MovieUpdateRequest(
        id=id,
        title=data.get('title'),
        year=data.get('year'),
        genre=data.get('genre'),
        description=data.get('description'),
        actors=data.get('actors', [])
    )
    response = movie_stub.UpdateMovie(movie_update_request)
    return jsonify({
        "id": response.id,
        "title": response.title,
        "year": response.year,
        "genre": response.genre,
        "description": response.description,
        "actors": list(response.actors)  # Convert to Python list
    }), 200

# Route to delete a movie via gRPC
@movie_routes.route('/<int:id>', methods=['DELETE'])
def delete_movie(id):
    movie_request = movies_pb2.MovieRequest(id=id)
    response = movie_stub.DeleteMovie(movie_request)
    return jsonify({"message": response.message}), 200

# Route to search movies via gRPC
@movie_routes.route('/search', methods=['GET'])
def search_movies():
    query = request.args.get('query', '')
    search_request = movies_pb2.SearchRequest(query=query)
    response = movie_stub.SearchMovies(search_request)
    movies = [{
        "id": movie.id,
        "title": movie.title,
        "year": movie.year,
        "genre": movie.genre,
        "description": movie.description,
        "actors": list(movie.actors)  # Convert to Python list
    } for movie in response.movies]
    return jsonify(movies), 200